src/slack_api.jl

```diff
@@ -48,16 +49,37 @@ function upload_file(local_path::AbstractString; extra_body=Dict())
         @debug "Uploading slack file" api local_path
     end
 
-    headers = ["Authorization" => "Bearer $(token)"]
+    headers = []
+
+    len_upload = string(stat(local_path).size)
     response = @maybecatch begin
-        open(local_path, "r") do file
-            body = HTTP.Form(vcat(collect(extra_body), ["file" => file]))
-            response = @mock HTTP.post(api, headers, body)
-            return JSON3.read(response.body)
-        end
+        body = HTTP.Form(vcat(collect(extra_body), ["token" => token, "filename" => local_path, "length" => len_upload]))
+        response = @mock HTTP.post(api, headers, body)
+        JSON3.read(response.body)
     end "Error when attempting to upload file to Slack"
 
-    response === nothing && return nothing
+    if haskey(response, :ok) === true && response[:ok] != true
+        return @maybecatch begin
+            err = haskey(response, :error) ? string(response.error) :
+                  "No error field returned"
+            throw(SlackError(err))
+        end "Error reported by Slack API"
+    end
+
+    upload_url = response[:upload_url]
+    file_data = read(local_path)
+    upload_response = @maybecatch begin
+        @mock HTTP.post(upload_url, [], file_data)
+    end "Error when attempting to upload file to Slack"
+
+    file_id = response[:file_id]
```

Comment:
It's been awhile, but I remember a key tenet of the package was that it should impossible to throw when errors are turned off. So since we can't necessarily trust the slack api to always return the right responses, let's check them more safely:
```suggestion
    upload_url = get(response, :upload_url, nothing)
    file_id = get(response, :file_id, nothing)
    if upload_url === nothing || file_id === nothing
        @maybe_catch begin
            throw(SlackError("Unexpected error: response missing `upload_url` or `file_id` fields"))
        end "Error when parsing Slack API response"
    end

   upload_response = @maybecatch begin
        @mock HTTP.post(upload_url, [], read(local_path))
    end "Error when attempting to upload file to Slack"

```

Reply:
Sounds good to me. I was also wondering if there should be checking on the upload response. upload_response.status should be 200 if the upload is successful.